api/explorer/api/ingest.py

```diff
@@ -210,30 +210,61 @@ def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
         )
 
     with ThreadPoolExecutor() as ex:
-        click.echo("Writing packages")
+        # Write nodes to the Gremlin graph
+        click.echo("Writing nodes to the Gremlin graph...")
         with click.progressbar(length=len(packages)) as bar:
             futures = [ex.submit(write_package, p) for p in packages]
             for f in futures:
                 f.add_done_callback(lambda _x: bar.update(1))
             wait(futures)
-            exceptions = [f.exception() for f in futures if f.exception()]
-            if exceptions:
-                # Raise a single exception after all tasks have completed
-                raise Exception(
-                    "Exceptions occurred while writing packages."
-                ) from exceptions[0]
-
-        click.echo("Writing edges")
+            # Collect the packages which are not ingested successfully
+            failed_packages = [
+                packages[futures.index(f)] for f in futures if f.exception()
+            ]
+            failed_packages_copy = failed_packages.copy()
+            if failed_packages_copy:
+                click.echo("\nRetrying to write nodes...")
+                for i in range(len(failed_packages_copy)):
+                    try:
+                        graph.insert_unique_vertex(failed_packages_copy[i], g_factory())
+                        failed_packages.remove(failed_packages_copy[i])
```

Comment:
If you don't want to juggle with indicies because of that, you can build another set and later compute the difference.
https://docs.python.org/3/library/stdtypes.html#frozenset.difference